### PR TITLE
[docs] Add end-to-end profiling example for Granite on Spyre

### LINKED ISSUE #1948: [docs] Add end-to-end profiling worked example (Granite via FMS)
Follow-up to #1746.

Add `docs/source/user_guide/profiling/end_to_end_example.md` — a runnable, end-to-end profiling example for a Granite-class model on Spyre, demonstrating today's tooling (`torch.profiler` + `kineto-spyre` + `aiu-smi` + `aiu-trace-analyzer`) on a real workload. Currently the profiling section covers each tool in isolation; readers don't have a single page that shows the full workflow on something representative.

## Public packages the example will use

- [`foundation-model-stack/foundation-model-stack`](https://github.com/foundation-model-stack/foundation-model-stack) — `fms`
- [`foundation-model-stack/aiu-fms-testing-utils`](https://github.com/foundation-model-stack/aiu-fms-testing-utils) — `aiu-fms-testing-utils`
- [`IBM/kineto-spyre`](https://github.com/IBM/kineto-spyre) — already covered in `pytorch_profiler.md`
- [`IBM/aiu-trace-analyzer`](https://github.com/IBM/aiu-trace-analyzer) — already covered in `trace_analysis.md`

## Patterns to demonstrate

- **Loading path.** Use `fms.models.get_model(...)` + `aiu-fms-testing-utils` rather than HuggingFace `AutoModelForCausalLM` — that's the path that actually runs end-to-end on Spyre today.
- **Warmup-discard recipe.** Run iteration 1 outside the timed loop (triggers `torch.compile` + kernel codegen); time iterations 2..N.
- **Dual timing.** Wall-clock via `time.perf_counter()` AND profiler-derived CPU time via `sum(e.self_cpu_time_total for e in profiler.events()) / active_runs`. Treat as orthogonal signals: wall-clock is what the user feels; CPU time is host-side overhead.
- **`tensorboard_trace_handler(log_dir)` over `export_chrome_trace`.** Per-step files, opens in TensorBoard *and* Chrome/Perfetto.
- **Two canonical shape regimes.** Prefill `[1, 512, 4096]` and decode `[4, 1, 4096]` — the standard inference-perf measurement points.
- **Minimal-repro trick.** Slice `model.base_model.layers = model.base_model.layers[:1]` to isolate per-layer perf when full-model traces are too noisy.
- **Forward-compat gate.** `if torch_spyre.profiler.is_available(): ... else: ...` so the example evolves cleanly when RFC 0601 lands.
- **Real numbers for calibration.** Include a few measured deltas (e.g. matmul / MLP / attention on the two shape regimes) so readers know what "good" looks like.

## Out of scope for this issue

- HuggingFace `AutoModelForCausalLM` path — defer until full Granite is supported through that path on Spyre.
- Any prep scripts that aren't public — the example must rest on a public Granite checkpoint and the public packages above.

## Validation requirement

The example must be **smoke-tested on Spyre hardware** before merge — it will be a fresh write, not a copy of any existing tested script, so drift is possible. Need to:

1. Run the example as written.
2. Verify it produces a valid Chrome trace and meaningful `aiu-smi` samples.
3. Confirm the cited "what to expect" numbers are within an order of magnitude of reproduction.

## Acceptance criteria

- [ ] New page `docs/source/user_guide/profiling/end_to_end_example.md` added to the toctree in `profiling/index.md`.
- [ ] All commands runnable on a public-package install.
- [ ] Smoke-tested end-to-end on Spyre hardware.
- [ ] `make -C docs html` clean.
- [ ] Cross-linked from `pytorch_profiler.md`, `device_monitoring.md`, and `performance_analysis_methodology.md`.

### PR DESCRIPTION
**Pull request by kaoutar55**
## Summary

Adds `docs/source/user_guide/profiling/end_to_end_example.md` — a runnable walkthrough that glues the existing per-tool profiling pages (`pytorch_profiler`, `device_monitoring`, `trace_analysis`, `performance_analysis_methodology`) into one end-to-end recipe for capturing a `torch.profiler` trace of a Granite-class model on Spyre, paired with `aiu-smi` device telemetry.

Closes #1948.

## Test plan

- [x] `make -C docs html`
- [ ] Validate the example end-to-end on Spyre hardware

### CODE REVIEW

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        wall_clock_ms.append((time.perf_counter() - t0) * 1000)
+        prof.step()
+
+# 6. Two timing signals: wall-clock (what the user feels) and
+#    profiler-derived CPU time (host-side overhead). The gap between
+#    them ≈ device-side work.
+cpu_per_run_ms = sum(e.self_cpu_time_total for e in prof.events()) / 1000 / N_RUNS
+
+print(
+    prof.key_averages()
```
Comment (ppnaik1890) [inline docs/source/user_guide/profiling/end_to_end_example.md]: print(prof.key_averages().table(sort_by="device_time_total", row_limit=10))
will not need the replace now since the registered  decide will be spyre now

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+The `.replace("CUDA", "AIU")` is a cosmetic workaround — the
+profiler's column category is still named after CUDA upstream; native
+renaming is on the roadmap. See [PyTorch Profiler](pytorch_profiler.md).
+
+## Run with telemetry alongside
+
+In one terminal:
+
+```bash
+aiu-smi dmon | tee /tmp/aiu-smi.log
```
Comment (ppnaik1890) [inline docs/source/user_guide/profiling/end_to_end_example.md:180]: env for senlib config file needs to be set before aiu-smi

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+python profile_granite.py
+```
+
+`aiu-smi dmon` samples once a second and streams power, temperature,
+PT-array utilization, device-memory and PCIe bandwidth. Pair its
+timestamps with the trace timeline to attribute idle gaps to either
+host-side work or device-side stalls. See
+[Device monitoring](device_monitoring.md).
+
+## Inspect the trace
```
Comment (ppnaik1890) [inline docs/source/user_guide/profiling/end_to_end_example.md]: this section can go before aiu-smi since it is related to profiler

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+## Forward compatibility
+
+Once [RFC 0601][rfc-0601] lands, the in-tree `torch_spyre.profiler`
+API will subsume the explicit `torch.profiler` block above. You can
+gate the script today so it picks up the in-tree path automatically
+when it ships:
+
+```python
+import torch_spyre.profiler
```
Comment (ppnaik1890) [inline docs/source/user_guide/profiling/end_to_end_example.md]: since we deprioritized the context manager issue should we remove this section

**Review by ppnaik1890 — DISMISSED** [review]

**Comment by kaoutar55** [comment]
Thanks @ppnaik1890. feedback addressed.

**Comment by kaoutar55** [comment]
we can merge this and add the example/screen shots in a follow up PR after @WinnieImafidon1 does her testing.

**Review by dgrove-oss — DISMISSED** [review]
lgtm

File: docs/source/user_guide/profiling/end_to_end_example.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+explicitly. Once [RFC 0601][rfc-0601] lands, the in-tree
+`torch_spyre.profiler` API will replace this glue and the script will
+shrink. This page will be revised to the in-tree API once it ships.
+:::
+
+## What you need
+
+| Piece | Source | Sample install (verify against the upstream README) |
+|---|---|---|
+| `foundation-model-stack` (`fms`) | [github.com/foundation-model-stack/foundation-model-stack][fms] | `pip install ibm-fms` |
```
Comment (ppnaik1890) [inline docs/source/user_guide/profiling/end_to_end_example.md]: The current support is in `eager_spyre` for both fms and aiu-fms-testing-utils

**Review by ppnaik1890 — COMMENTED** [review]
minor comment else lgtm

**Review by dgrove-oss — APPROVED** [review]

**Review by ppnaik1890 — APPROVED** [review]
